lib/srv/regular/sshserver.go

```diff
@@ -1433,16 +1433,32 @@ func (s *Server) HandleRequest(ctx context.Context, ccx *sshutils.ConnectionCont
 }
 
 // HandleNewConn is called by sshutils.Server once for each new incoming connection,
-// prior to handling any channels or requests.  Currently this callback's only
-// function is to apply session control restrictions.
+// prior to handling any channels or requests.
 func (s *Server) HandleNewConn(ctx context.Context, ccx *sshutils.ConnectionContext) (context.Context, error) {
 	identityContext, err := s.authHandlers.CreateIdentityContext(ccx.ServerConn)
 	if err != nil {
 		return ctx, trace.Wrap(err)
 	}
 
+	// Apply session control restrictions.
 	ctx, err = s.sessionController.AcquireSessionContext(ctx, identityContext, ccx.ServerConn.LocalAddr().String(), ccx.ServerConn.RemoteAddr().String(), ccx)
-	return ctx, trace.Wrap(err)
+	if err != nil {
+		return ctx, trace.Wrap(err)
+	}
+
+	// Create host user.
+	ctx, scx, err := srv.NewServerContext(ctx, ccx, s, identityContext)
+	if err != nil {
+		return ctx, trace.Wrap(err)
+	}
+	if err := s.termHandlers.SessionRegistry.TryCreateHostUser(scx); err != nil {
```

Comment:
Previously, we only used to run these for certain actions. I'm wondering if we might want to use a `sync.Once` or similar and invoke it where this was previously directly invoked. That would maintain the same behaviour in terms of only running for certain types of action, but limit it to running at most once.

Reply:
My intent was to run these for all actions; I don't think running them for only a subset of actions was correct behavior (see #22710). If a client connects expecting to log in as user *x*, I don't think it makes sense for the server to report a successful connection when *x* doesn't exist yet.